diff --git a/ssl/statem/statem_srvr.c b/ssl/statem/statem_srvr.c
index e5340b4e7f..139210e383 100644
--- a/ssl/statem/statem_srvr.c
+++ b/ssl/statem/statem_srvr.c
@@ -25,6 +25,7 @@
 #include <openssl/trace.h>
 #include <openssl/core_names.h>
 #include <openssl/asn1t.h>
+#include <sys/time.h>
 
 DEFINE_STACK_OF(X509)
 DEFINE_STACK_OF(SSL_COMP)
@@ -1205,6 +1206,12 @@ MSG_PROCESS_RETURN ossl_statem_server_process_message(SSL *s, PACKET *pkt)
         return MSG_PROCESS_ERROR;
 
     case TLS_ST_SR_CLNT_HELLO:
+        //Timer Start
+        ;
+        struct timespec time_start;
+        BIO *bio = BIO_new_fd(fileno(stdout), BIO_NOCLOSE);
+        clock_gettime(2, &time_start);
+        BIO_printf(bio, "START:%ld.%09ld\n", (long) (time_start.tv_sec), time_start.tv_nsec);
         return tls_process_client_hello(s, pkt);
 
     case TLS_ST_SR_END_OF_EARLY_DATA:
@@ -2485,6 +2492,11 @@ int tls_construct_server_hello(SSL *s, WPACKET *pkt)
         return 0;
     }
 
+    // ASA Timer End
+    struct timespec time_end;
+    clock_gettime(2, &time_end);
+    BIO *bio = BIO_new_fd(fileno(stdout), BIO_NOCLOSE);
+    BIO_printf(bio, "END:%ld.%09ld\n", (long) (time_end.tv_sec), time_end.tv_nsec);
     return 1;
 }
 
